File: python/src/bleepstore/handlers/bucket.py
"""Bucket-level S3 request handlers for BleepStore.

Implements the 7 bucket operations:
    - ListBuckets (GET /)
    - CreateBucket (PUT /{bucket})
    - DeleteBucket (DELETE /{bucket})
    - HeadBucket (HEAD /{bucket})
    - GetBucketLocation (GET /{bucket}?location)
    - GetBucketAcl (GET /{bucket}?acl)
    - PutBucketAcl (PUT /{bucket}?acl)
"""

import hashlib
import logging
from xml.etree import ElementTree

from fastapi import FastAPI, Request, Response

from bleepstore.errors import (
    BucketNotEmpty,
    InvalidBucketName,
    MalformedXML,
    NoSuchBucket,
)
from bleepstore.handlers.acl import (
    acl_from_json,
    acl_to_json,
    build_default_acl,
    parse_canned_acl,
    render_acl_xml,
)
from bleepstore.validation import validate_bucket_name
from bleepstore.xml_utils import (
    render_list_buckets,
    render_location_constraint,
    xml_response,
)

logger = logging.getLogger(__name__)


def _derive_owner_id(access_key: str) -> str:
    """Derive a canonical owner ID from an access key.

    Uses SHA-256 hash of the access key, truncated to 32 characters.

    Args:
        access_key: The AWS access key string.

    Returns:
        A 32-character hex string owner ID.
    """
    return hashlib.sha256(access_key.encode()).hexdigest()[:32]


class BucketHandler:
    """Handles S3 bucket operations.

    All handlers access metadata and config from ``app.state``.

    Attributes:
        app: The parent FastAPI application.
    """

    def __init__(self, app: FastAPI) -> None:
        """Initialize the bucket handler.

        Args:
            app: The FastAPI application instance.
        """
        self.app = app

    @property
    def metadata(self):
        """Shortcut to the metadata store on app.state."""
        return self.app.state.metadata

    @property
    def config(self):
        """Shortcut to the BleepStoreConfig on app.state."""
        return self.app.state.config

    async def list_buckets(self, request: Request) -> Response:
        """List all buckets owned by the authenticated user.

        Implements: GET /

        Args:
            request: The incoming HTTP request.

        Returns:
            XML response containing the bucket list.
        """
        access_key = self.config.auth.access_key
        owner_id = _derive_owner_id(access_key)

        buckets = await self.metadata.list_buckets()

        xml = render_list_buckets(
            owner_id=owner_id,
            owner_display_name=access_key,
            buckets=buckets,
        )
        return xml_response(xml, status=200)

    async def create_bucket(self, request: Request, bucket: str) -> Response:
        """Create a new bucket.

        Implements: PUT /{bucket}

        Idempotency: if the bucket already exists and is owned by the caller,
        returns 200 (us-east-1 behavior).

        Args:
            request: The incoming HTTP request.
            bucket: The bucket name from the URL path.

        Returns:
            Empty 200 response on success with Location header.
        """
        # Validate bucket name
        try:
            validate_bucket_name(bucket)
        except InvalidBucketName:
            raise

        # Determine the region from the request body (CreateBucketConfiguration)
        region = self.config.server.region
        body = await request.body()
        if body:
            try:
                root = ElementTree.fromstring(body)
                # Handle namespace
                ns = "{http://s3.amazonaws.com/doc/2006-03-01/}"
                loc_elem = root.find(f"{ns}LocationConstraint")
                if loc_elem is None:
                    # Try without namespace
                    loc_elem = root.find("LocationConstraint")
                if loc_elem is not None and loc_elem.text:
                    region = loc_elem.text.strip()
            except ElementTree.ParseError:
                raise MalformedXML()

        # Derive owner from access key
        access_key = self.config.auth.access_key
        owner_id = _derive_owner_id(access_key)
        owner_display = access_key

        # Check if bucket already exists
        existing = await self.metadata.get_bucket(bucket)
        if existing is not None:
            # us-east-1 behavior: if owned by caller, return 200
            # Since we have a single-user system, any existing bucket
            # is "owned by you".
            return Response(
                status_code=200,
                headers={"Location": f"/{bucket}"},
            )

        # Build default ACL
        acl = build_default_acl(owner_id, owner_display)
        acl_json = acl_to_json(acl)

        # Determine canned ACL from header
        canned_acl = request.headers.get("x-amz-acl")
        if canned_acl:
            acl = parse_canned_acl(canned_acl, owner_id, owner_display)
            acl_json = acl_to_json(acl)

        # Create the bucket in metadata store
        await self.metadata.create_bucket(
            bucket=bucket,
            region=region,
            owner_id=owner_id,
            owner_display=owner_display,
            acl=acl_json,
        )

        return Response(
            status_code=200,
            headers={"Location": f"/{bucket}"},
        )

    async def delete_bucket(self, request: Request, bucket: str) -> Response:
        """Delete an existing empty bucket.

        Implements: DELETE /{bucket}

        Preconditions:
            - Bucket must exist (NoSuchBucket if not).
            - Bucket must be empty (BucketNotEmpty if objects exist).

        Args:
            request: The incoming HTTP request.
            bucket: The bucket name from the URL path.

        Returns:
            204 No Content on success.
        """
        # Check if bucket exists
        exists = await self.metadata.bucket_exists(bucket)
        if not exists:
            raise NoSuchBucket(bucket)

        # Check if bucket has any objects
        obj_count = await self.metadata.count_objects(bucket)
        if obj_count > 0:
            raise BucketNotEmpty(bucket)

        # Delete the bucket
        await self.metadata.delete_bucket(bucket)

        return Response(status_code=204)

    async def head_bucket(self, request: Request, bucket: str) -> Response:
        """Check if a bucket exists and the caller has access.

        Implements: HEAD /{bucket}

        HEAD responses never have a body.

        Args:
            request: The incoming HTTP request.
            bucket: The bucket name from the URL path.

        Returns:
            200 OK if the bucket exists, with x-amz-bucket-region header.
        """
        bucket_meta = await self.metadata.get_bucket(bucket)
        if bucket_meta is None:
            raise NoSuchBucket(bucket)

        region = bucket_meta.get("region", "us-east-1")
        return Response(
            status_code=200,
            headers={"x-amz-bucket-region": region},
        )

    async def get_bucket_location(
        self, request: Request, bucket: str
    ) -> Response:
        """Return the region where the bucket resides.

        Implements: GET /{bucket}?location

        us-east-1 quirk: returns empty ``<LocationConstraint/>`` instead of
        the string ``us-east-1``.

        Args:
            request: The incoming HTTP request.
            bucket: The bucket name from the URL path.

        Returns:
            XML response with the bucket location constraint.
        """
        bucket_meta = await self.metadata.get_bucket(bucket)
        if bucket_meta is None:
            raise NoSuchBucket(bucket)

        region = bucket_meta.get("region", "us-east-1")
        xml = render_location_constraint(region)
        return xml_response(xml, status=200)

    async def get_bucket_acl(
        self, request: Request, bucket: str
    ) -> Response:
        """Return the access control list for a bucket.

        Implements: GET /{bucket}?acl

        Args:
            request: The incoming HTTP request.
            bucket: The bucket name from the URL path.

        Returns:
            XML response with the bucket ACL.
        """
        bucket_meta = await self.metadata.get_bucket(bucket)
        if bucket_meta is None:
            raise NoSuchBucket(bucket)

        acl_json_str = bucket_meta.get("acl", "{}")
        acl = acl_from_json(acl_json_str)

        # If the ACL has no owner info, fill it in from bucket metadata
        if not acl.get("owner", {}).get("id"):
            owner_id = bucket_meta.get("owner_id", "")
            owner_display = bucket_meta.get("owner_display", "")
            if not acl.get("grants"):
                # Build a default ACL
                acl = build_default_acl(owner_id, owner_display)
            else:
                acl["owner"] = {"id": owner_id, "display_name": owner_display}

        xml = render_acl_xml(acl)
        return xml_response(xml, status=200)

    async def put_bucket_acl(
        self, request: Request, bucket: str
    ) -> Response:
        """Set the access control list for a bucket.

        Implements: PUT /{bucket}?acl

        Supports three mutually exclusive modes:
            1. Canned ACL via x-amz-acl header
            2. XML body with full AccessControlPolicy
            3. (x-amz-grant-* headers - simplified support)

        Args:
            request: The incoming HTTP request.
            bucket: The bucket name from the URL path.

        Returns:
            Empty 200 response on success.
        """
        bucket_meta = await self.metadata.get_bucket(bucket)
        if bucket_meta is None:
            raise NoSuchBucket(bucket)

        owner_id = bucket_meta.get("owner_id", "")
        owner_display = bucket_meta.get("owner_display", "")

        # Check for canned ACL header
        canned_acl = request.headers.get("x-amz-acl")
        if canned_acl:
            try:
                acl = parse_canned_acl(canned_acl, owner_id, owner_display)
            except ValueError:
                raise InvalidBucketName()  # S3 returns InvalidArgument for bad ACL names
            acl_json = acl_to_json(acl)
            await self.metadata.update_bucket_acl(bucket, acl_json)
            return Response(status_code=200)

        # Check for XML body
        body = await request.body()
        if body:
            try:
                root = ElementTree.fromstring(body)
                # Parse the AccessControlPolicy XML into our ACL format
                acl = _parse_acl_xml(root, owner_id, owner_display)
                acl_json = acl_to_json(acl)
                await self.metadata.update_bucket_acl(bucket, acl_json)
                return Response(status_code=200)
            except ElementTree.ParseError:
                raise MalformedXML()

        # No ACL specified -- default to private
        acl = build_default_acl(owner_id, owner_display)
        acl_json = acl_to_json(acl)
        await self.metadata.update_bucket_acl(bucket, acl_json)
        return Response(status_code=200)


def _find_elem(
    parent: ElementTree.Element, ns_name: str, bare_name: str
) -> ElementTree.Element | None:
    """Find a child element, trying namespaced name first, then bare name.

    Uses explicit ``is not None`` checks to avoid ElementTree's deprecated
    truth-value testing of elements.

    Args:
        parent: The parent XML element to search.
        ns_name: The namespace-qualified element name.
        bare_name: The element name without namespace.

    Returns:
        The found element, or None.
    """
    elem = parent.find(ns_name)
    if elem is not None:
        return elem
    return parent.find(bare_name)


def _parse_acl_xml(
    root: ElementTree.Element,
    default_owner_id: str,
    default_owner_display: str,
) -> dict:
    """Parse an AccessControlPolicy XML element into an ACL dict.

    Args:
        root: The parsed XML root element.
        default_owner_id: Fallback owner ID if not in XML.
        default_owner_display: Fallback display name if not in XML.

    Returns:
        An ACL dict with 'owner' and 'grants' keys.
    """
    ns = "{http://s3.amazonaws.com/doc/2006-03-01/}"

    # Parse owner
    owner_elem = _find_elem(root, f"{ns}Owner", "Owner")
    owner_id = default_owner_id
    owner_display = default_owner_display
    if owner_elem is not None:
        id_elem = _find_elem(owner_elem, f"{ns}ID", "ID")
        if id_elem is not None and id_elem.text:
            owner_id = id_elem.text
        dn_elem = _find_elem(owner_elem, f"{ns}DisplayName", "DisplayName")
        if dn_elem is not None and dn_elem.text:
            owner_display = dn_elem.text

    # Parse grants
    grants = []
    acl_elem = _find_elem(root, f"{ns}AccessControlList", "AccessControlList")
    if acl_elem is not None:
        for grant_elem in (
            acl_elem.findall(f"{ns}Grant") + acl_elem.findall("Grant")
        ):
            grantee_elem = _find_elem(
                grant_elem, f"{ns}Grantee", "Grantee"
            )
            perm_elem = _find_elem(
                grant_elem, f"{ns}Permission", "Permission"
            )
            if grantee_elem is None or perm_elem is None:
                continue

            permission = perm_elem.text or ""

            # Determine grantee type from xsi:type attribute
            xsi_type = grantee_elem.get(
                "{http://www.w3.org/2001/XMLSchema-instance}type", ""
            )

            if xsi_type == "Group" or grantee_elem.find(f"{ns}URI") is not None:
                uri_elem = _find_elem(grantee_elem, f"{ns}URI", "URI")
                uri = uri_elem.text if uri_elem is not None else ""
                grants.append({
                    "grantee": {"type": "Group", "uri": uri},
                    "permission": permission,
                })
            else:
                # CanonicalUser
                g_id_elem = _find_elem(grantee_elem, f"{ns}ID", "ID")
                g_dn_elem = _find_elem(
                    grantee_elem, f"{ns}DisplayName", "DisplayName"
                )
                g_id = g_id_elem.text if g_id_elem is not None else ""
                g_dn = g_dn_elem.text if g_dn_elem is not None else ""
                grants.append({
                    "grantee": {
                        "type": "CanonicalUser",
                        "id": g_id,
                        "display_name": g_dn,
                    },
                    "permission": permission,
                })

    return {
        "owner": {"id": owner_id, "display_name": owner_display},
        "grants": grants,
    }

File: golang/internal/handlers/bucket.go
// Package handlers implements HTTP request handlers for S3-compatible API operations.
package handlers

import (
	"context"
	"encoding/xml"
	"io"
	"log/slog"
	"net/http"
	"strings"
	"time"

	s3err "github.com/bleepstore/bleepstore/internal/errors"
	"github.com/bleepstore/bleepstore/internal/metadata"
	"github.com/bleepstore/bleepstore/internal/storage"
	"github.com/bleepstore/bleepstore/internal/xmlutil"
)

// BucketHandler contains handlers for S3 bucket-level operations.
type BucketHandler struct {
	meta         metadata.MetadataStore
	store        storage.StorageBackend
	ownerID      string
	ownerDisplay string
	region       string
}

// NewBucketHandler creates a new BucketHandler with the given dependencies.
func NewBucketHandler(meta metadata.MetadataStore, store storage.StorageBackend, ownerID, ownerDisplay, region string) *BucketHandler {
	return &BucketHandler{
		meta:         meta,
		store:        store,
		ownerID:      ownerID,
		ownerDisplay: ownerDisplay,
		region:       region,
	}
}

// ListBuckets handles GET / and returns a list of all buckets owned by the
// authenticated sender of the request.
func (h *BucketHandler) ListBuckets(w http.ResponseWriter, r *http.Request) {
	if h.meta == nil {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	ctx := r.Context()

	buckets, err := h.meta.ListBuckets(ctx, h.ownerID)
	if err != nil {
		slog.Error("ListBuckets error", "error", err)
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	var xmlBuckets []xmlutil.Bucket
	for _, b := range buckets {
		xmlBuckets = append(xmlBuckets, xmlutil.Bucket{
			Name:         b.Name,
			CreationDate: xmlutil.FormatTimeS3(b.CreatedAt),
		})
	}

	result := &xmlutil.ListAllMyBucketsResult{
		Owner: xmlutil.Owner{
			ID:          h.ownerID,
			DisplayName: h.ownerDisplay,
		},
		Buckets: xmlBuckets,
	}

	xmlutil.RenderListBuckets(w, result)
}

// CreateBucket handles PUT /{bucket} and creates a new bucket with the
// specified name.
func (h *BucketHandler) CreateBucket(w http.ResponseWriter, r *http.Request) {
	if h.meta == nil || h.store == nil {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	ctx := r.Context()
	bucketName := extractBucketName(r)

	// Validate bucket name.
	if errMsg := validateBucketName(bucketName); errMsg != "" {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInvalidBucketName)
		return
	}

	// Parse optional canned ACL from header.
	cannedACL := r.Header.Get("x-amz-acl")

	// Build ACL: if canned ACL specified, use it; otherwise default to private.
	acp := parseCannedACL(cannedACL, h.ownerID, h.ownerDisplay)
	aclJSON := aclToJSON(acp)

	// Determine region from request body (CreateBucketConfiguration) or config.
	region := h.region
	if r.ContentLength > 0 || r.Header.Get("Content-Length") != "" {
		body, err := io.ReadAll(io.LimitReader(r.Body, 1<<20)) // 1 MB max
		if err == nil && len(body) > 0 {
			region = parseCreateBucketRegion(body, h.region)
		}
	}

	// Check if bucket already exists.
	existing, err := h.meta.GetBucket(ctx, bucketName)
	if err != nil {
		slog.Error("CreateBucket GetBucket error", "error", err)
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	if existing != nil {
		// Bucket already exists.
		if existing.OwnerID == h.ownerID {
			// us-east-1 behavior: return 200 OK (BucketAlreadyOwnedByYou).
			w.Header().Set("Location", "/"+bucketName)
			w.WriteHeader(http.StatusOK)
			return
		}
		// Bucket owned by someone else.
		xmlutil.WriteErrorResponse(w, r, s3err.ErrBucketAlreadyExists)
		return
	}

	// Create bucket record in metadata store.
	record := &metadata.BucketRecord{
		Name:         bucketName,
		Region:       region,
		OwnerID:      h.ownerID,
		OwnerDisplay: h.ownerDisplay,
		ACL:          aclJSON,
		CreatedAt:    time.Now().UTC(),
	}

	if err := h.meta.CreateBucket(ctx, record); err != nil {
		// Handle race condition: bucket was created between our check and insert.
		if strings.Contains(err.Error(), "already exists") {
			w.Header().Set("Location", "/"+bucketName)
			w.WriteHeader(http.StatusOK)
			return
		}
		slog.Error("CreateBucket metadata error", "error", err)
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	// Create the bucket directory in the storage backend.
	if err := h.store.CreateBucket(ctx, bucketName); err != nil {
		slog.Error("CreateBucket storage error", "error", err)
		// Best effort: metadata is created, storage directory failed.
		// Log but don't fail -- the directory will be created on first object write.
	}

	w.Header().Set("Location", "/"+bucketName)
	w.WriteHeader(http.StatusOK)
}

// DeleteBucket handles DELETE /{bucket} and removes the specified bucket.
// The bucket must be empty before it can be deleted.
func (h *BucketHandler) DeleteBucket(w http.ResponseWriter, r *http.Request) {
	if h.meta == nil || h.store == nil {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	ctx := r.Context()
	bucketName := extractBucketName(r)

	// Delete from metadata store (validates existence and emptiness).
	if err := h.meta.DeleteBucket(ctx, bucketName); err != nil {
		if strings.Contains(err.Error(), "not found") {
			xmlutil.WriteErrorResponse(w, r, s3err.ErrNoSuchBucket)
			return
		}
		if strings.Contains(err.Error(), "not empty") {
			xmlutil.WriteErrorResponse(w, r, s3err.ErrBucketNotEmpty)
			return
		}
		slog.Error("DeleteBucket error", "error", err)
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	// Remove bucket directory from storage backend (best effort).
	if err := h.store.DeleteBucket(ctx, bucketName); err != nil {
		slog.Error("DeleteBucket storage cleanup error", "error", err)
	}

	w.WriteHeader(http.StatusNoContent)
}

// HeadBucket handles HEAD /{bucket} and checks whether the specified bucket
// exists and is accessible.
func (h *BucketHandler) HeadBucket(w http.ResponseWriter, r *http.Request) {
	if h.meta == nil {
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	ctx := r.Context()
	bucketName := extractBucketName(r)

	bucket, err := h.meta.GetBucket(ctx, bucketName)
	if err != nil {
		slog.Error("HeadBucket error", "error", err)
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	if bucket == nil {
		// HEAD requests: no body, status code only.
		w.WriteHeader(http.StatusNotFound)
		return
	}

	w.Header().Set("x-amz-bucket-region", bucket.Region)
	w.WriteHeader(http.StatusOK)
}

// GetBucketLocation handles GET /{bucket}?location and returns the region
// constraint for the specified bucket.
func (h *BucketHandler) GetBucketLocation(w http.ResponseWriter, r *http.Request) {
	if h.meta == nil {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	ctx := r.Context()
	bucketName := extractBucketName(r)

	bucket, err := h.meta.GetBucket(ctx, bucketName)
	if err != nil {
		slog.Error("GetBucketLocation error", "error", err)
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	if bucket == nil {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrNoSuchBucket)
		return
	}

	// us-east-1 quirk: return empty LocationConstraint (effectively null).
	location := bucket.Region
	if location == "us-east-1" {
		location = ""
	}

	xmlutil.RenderLocationConstraint(w, location)
}

// GetBucketAcl handles GET /{bucket}?acl and returns the access control list
// for the specified bucket.
func (h *BucketHandler) GetBucketAcl(w http.ResponseWriter, r *http.Request) {
	if h.meta == nil {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	ctx := r.Context()
	bucketName := extractBucketName(r)

	bucket, err := h.meta.GetBucket(ctx, bucketName)
	if err != nil {
		slog.Error("GetBucketAcl error", "error", err)
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	if bucket == nil {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrNoSuchBucket)
		return
	}

	// Parse ACL from stored JSON.
	acp := aclFromJSON(bucket.ACL)
	if acp == nil {
		// No ACL stored: return default private ACL.
		acp = parseCannedACL("private", bucket.OwnerID, bucket.OwnerDisplay)
	}

	// Ensure Owner is set correctly.
	acp.Owner = xmlutil.Owner{
		ID:          bucket.OwnerID,
		DisplayName: bucket.OwnerDisplay,
	}

	xmlutil.RenderAccessControlPolicy(w, acp)
}

// PutBucketAcl handles PUT /{bucket}?acl and sets the access control list
// for the specified bucket.
func (h *BucketHandler) PutBucketAcl(w http.ResponseWriter, r *http.Request) {
	if h.meta == nil {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	ctx := r.Context()
	bucketName := extractBucketName(r)

	// Verify bucket exists.
	bucket, err := h.meta.GetBucket(ctx, bucketName)
	if err != nil {
		slog.Error("PutBucketAcl error", "error", err)
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	if bucket == nil {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrNoSuchBucket)
		return
	}

	var acp *xmlutil.AccessControlPolicy

	// Three mutually exclusive modes:
	// 1. Canned ACL via x-amz-acl header
	// 2. Explicit grants via x-amz-grant-* headers
	// 3. XML body
	cannedACL := r.Header.Get("x-amz-acl")
	if cannedACL != "" {
		// Mode 1: Canned ACL.
		acp = parseCannedACL(cannedACL, bucket.OwnerID, bucket.OwnerDisplay)
	} else if r.ContentLength > 0 {
		// Mode 3: XML body.
		body, readErr := io.ReadAll(io.LimitReader(r.Body, 1<<20)) // 1 MB max
		if readErr != nil {
			xmlutil.WriteErrorResponse(w, r, s3err.ErrMalformedXML)
			return
		}
		acp = &xmlutil.AccessControlPolicy{}
		if xmlErr := xml.Unmarshal(body, acp); xmlErr != nil {
			xmlutil.WriteErrorResponse(w, r, s3err.ErrMalformedXML)
			return
		}
	} else {
		// No canned ACL and no body: default to private.
		acp = parseCannedACL("private", bucket.OwnerID, bucket.OwnerDisplay)
	}

	// Store the ACL.
	aclJSON := aclToJSON(acp)
	if err := h.meta.UpdateBucketAcl(ctx, bucketName, aclJSON); err != nil {
		slog.Error("PutBucketAcl update error", "error", err)
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return
	}

	w.WriteHeader(http.StatusOK)
}

// parseCreateBucketRegion parses a CreateBucketConfiguration XML body to
// extract the LocationConstraint value. Returns the default region if
// parsing fails or no LocationConstraint is specified.
func parseCreateBucketRegion(body []byte, defaultRegion string) string {
	type createBucketConfig struct {
		XMLName            xml.Name `xml:"CreateBucketConfiguration"`
		LocationConstraint string   `xml:"LocationConstraint"`
	}
	var config createBucketConfig
	if err := xml.Unmarshal(body, &config); err != nil {
		return defaultRegion
	}
	if config.LocationConstraint == "" {
		return defaultRegion
	}
	return config.LocationConstraint
}

// ensureBucketExists is a helper that checks for bucket existence and writes
// the appropriate error response if it does not exist. Returns the bucket
// record if found, nil otherwise.
func (h *BucketHandler) ensureBucketExists(w http.ResponseWriter, r *http.Request, ctx context.Context, bucketName string) *metadata.BucketRecord {
	bucket, err := h.meta.GetBucket(ctx, bucketName)
	if err != nil {
		slog.Error("ensureBucketExists error", "error", err)
		xmlutil.WriteErrorResponse(w, r, s3err.ErrInternalError)
		return nil
	}
	if bucket == nil {
		xmlutil.WriteErrorResponse(w, r, s3err.ErrNoSuchBucket)
		return nil
	}
	return bucket
}
